engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceAnywhereTest.java

```diff
@@ -0,0 +1,423 @@
+
+    // Then
+    Assertions.assertThat(
+            RecordingExporter.processInstanceRecords()
+                .withProcessInstanceKey(key)
+                .withElementType(BpmnElementType.PROCESS)
+                .withIntent(ProcessInstanceIntent.ELEMENT_COMPLETED)
+                .limit(1))
+        .hasSize(1);
+
+    Assertions.assertThat(
+            RecordingExporter.processInstanceRecords()
+                .withProcessInstanceKey(key)
+                .limitToProcessInstanceCompleted())
+        .extracting(record -> record.getValue().getBpmnElementType())
+        .doesNotContain(BpmnElementType.START_EVENT);
+
+    Assertions.assertThat(
+            RecordingExporter.processInstanceRecords()
+                .withProcessInstanceKey(key)
+                .limitToProcessInstanceCompleted())
+        .extracting(record -> record.getValue().getBpmnElementType(), Record::getIntent)
+        .containsSubsequence(
+            Tuple.tuple(BpmnElementType.PROCESS, ProcessInstanceIntent.ELEMENT_ACTIVATING),
+            Tuple.tuple(BpmnElementType.PROCESS, ProcessInstanceIntent.ELEMENT_ACTIVATED),
+            Tuple.tuple(BpmnElementType.SUB_PROCESS, ProcessInstanceIntent.ELEMENT_ACTIVATING),
+            Tuple.tuple(BpmnElementType.SUB_PROCESS, ProcessInstanceIntent.ELEMENT_ACTIVATED),
+            Tuple.tuple(BpmnElementType.SUB_PROCESS, ProcessInstanceIntent.ELEMENT_ACTIVATING),
+            Tuple.tuple(BpmnElementType.SUB_PROCESS, ProcessInstanceIntent.ELEMENT_ACTIVATED),
+            Tuple.tuple(BpmnElementType.SERVICE_TASK, ProcessInstanceIntent.ELEMENT_ACTIVATING),
+            Tuple.tuple(BpmnElementType.SERVICE_TASK, ProcessInstanceIntent.ELEMENT_ACTIVATED));
+  }
+
+  @Test
+  public void shouldCreateVariablesInProcessScope() {
+    // Given
+    ENGINE
+        .deployment()
+        .withXmlResource(
+            Bpmn.createExecutableProcess("process").startEvent().endEvent("end").done())
+        .deploy();
+
+    // When
+    final long key =
+        ENGINE
+            .processInstance()
+            .ofBpmnProcessId("process")
+            .withStartInstruction(newStartInstruction("end"))
+            .withVariable("variable", "value")
+            .create();
+
+    // Then
+    Assertions.assertThat(
+            RecordingExporter.processInstanceRecords()
+                .withProcessInstanceKey(key)
+                .withElementType(BpmnElementType.PROCESS)
+                .withIntent(ProcessInstanceIntent.ELEMENT_COMPLETED)
+                .limit(1))
+        .hasSize(1);
+
+    Assertions.assertThat(
+            RecordingExporter.variableRecords(VariableIntent.CREATED)
+                .withProcessInstanceKey(key)
+                .getFirst())
+        .extracting(Record::getValue)
+        .extracting(
+            VariableRecordValue::getScopeKey,
+            VariableRecordValue::getName,
+            VariableRecordValue::getValue)
+        .containsExactly(key, "variable", "\"value\"");
```

Comment:
:wrench: We could limit the stream to the process instance instead to verify that the variable is created only for the process instance and not more. For example:

```
RecordingExporter.
  records()
  .limitToProcessInstance(key)
  .variableRecords()
  // ...
```

Reply:
A limit like this doesn't work as `limitToProcessInstance` also filters on specific `ProcessInstanceIntents`, thus not returning the variable records. I don't see another way to limit, am I missing something?

Reply:
Not true. It limits the stream based on the process instance records of the given process instance key but it returns the variable records.

However, I just realized that it doesn't work because the variables are written before the process instance record of the process is written. 

But we could write it in the following way:

```
Assertions.assertThat(
            RecordingExporter.records()
                .limit(
                    // extract the condition into a method like limitToProcessInstance()
                    r ->
                        r.getKey() == key
                            && r.getIntent() == ProcessInstanceIntent.ELEMENT_COMPLETED)
                .variableRecords())
        .hasSize(1)
        .extracting(Record::getValue)
        .extracting(
            VariableRecordValue::getScopeKey,
            VariableRecordValue::getName,
            VariableRecordValue::getValue)
        .containsExactly(tuple(key, "variable", "\"value\""));
```